File: desktop/src/core/FileSystemEntries.cpp
#include "FileSystemEntries.h"
#include <iostream>

namespace ImageToolkit
{
    // --- Utility Method Implementations ---

    void FileSystemEntries::ensureDirectoryExists(const fs::path& filePath)
    {
        fs::path directory = filePath.parent_path();
        if (!directory.empty() && !fs::exists(directory))
        {
            try
            {
                fs::create_directories(directory);
                std::cout << "Created directory: '" << directory.string() << "'." << std::endl;
            }
            catch (const std::exception& e)
            {
                throw ImageToolException("Could not create directory: " + directory.string() + ". Reason: " + e.what());
            }
        }
    }

    fs::path FileSystemEntries::makeAbsolute(const fs::path& path)
    {
        return fs::absolute(path);
    }

    // --- Core Method Implementations ---

    bool FileSystemEntries::pathContains(const fs::path& parentPath, const fs::path& childPath)
    {
        try
        {
            fs::path parent = fs::canonical(parentPath);
            fs::path child = fs::canonical(childPath);

            // Iterate up from the child to see if we find the parent
            auto it = child.parent_path();
            while (!it.empty() && it != it.root_path()) {
                if (it == parent) {
                    return true;
                }
                it = it.parent_path();
            }
            // Check the root case
            return it == parent;
        }
        catch (const fs::filesystem_error&)
        {
            return false;
        }
    }

    int FileSystemEntries::deleteFilesByExtensions(fs::path directory, const std::vector<std::string>& extensions)
    {
        directory = makeAbsolute(directory);
        int deletedCount = 0;

        for (const auto& ext : extensions)
        {
            std::string targetExt = ext.find('.') == 0 ? ext : "." + ext;
            targetExt = to_lower(targetExt);

            // Use recursive_directory_iterator to match rglob
            for (const auto& entry : fs::recursive_directory_iterator(directory))
            {
                // FIX 1: Use fs::is_regular_file(entry) instead of entry.is_file()
                if (fs::is_regular_file(entry) && to_lower(entry.path().extension().string()) == targetExt)
                {
                    try
                    {
                        fs::remove(entry.path());
                        std::cout << "Deleted: " << entry.path().string() << std::endl;
                        deletedCount++;
                    }
                    catch (const std::exception& e)
                    {
                        std::cerr << "ERROR: Could not delete file " << entry.path().string() << ". Reason: " << e.what() << std::endl;
                    }
                }
            }
        }
        std::cout << "Deleted " << deletedCount << " files recursively." << std::endl;
        return deletedCount;
    }

    bool FileSystemEntries::deletePath(fs::path pathToDelete)
    {
        pathToDelete = makeAbsolute(pathToDelete);

        if (!fs::exists(pathToDelete))
        {
            std::cerr << "WARNING: specified path does not exist - did not delete '" << pathToDelete.string() << "'." << std::endl;
            return false;
        }

        try
        {
            if (fs::is_directory(pathToDelete))
            {
                fs::remove_all(pathToDelete); // Replaces shutil.rmtree
                std::cout << "Deleted directory: '" << pathToDelete.string() << "'." << std::endl;
            }
            else if (fs::is_regular_file(pathToDelete)) // Use fs::is_regular_file for consistency
            {
                fs::remove(pathToDelete); // Replaces os.remove
                std::cout << "Deleted file: '" << pathToDelete.string() << "'." << std::endl;
            }
            return true;
        }
        catch (const std::exception& e)
        {
            // FIX 2: Use << to stream C-style strings, not +
            std::cerr << "ERROR: Could not delete " << pathToDelete.string() << ". Reason: " << e.what() << std::endl;
            return false;
        }
    }

    std::vector<fs::path> FileSystemEntries::getFilesByExtension(fs::path directory, std::string extension, bool recursive)
    {
        directory = makeAbsolute(directory);
        std::vector<fs::path> files;

        // FIX 3: Use C++17-compatible check (std::string::starts_with is C++20)
        if (extension.empty() || extension[0] != '.') {
            extension = "." + extension;
        }
        extension = to_lower(extension);

        auto add_file_if_match = [&](const fs::directory_entry& entry) {
            // Use fs::is_regular_file for consistency
            if (fs::is_regular_file(entry) && to_lower(entry.path().extension().string()) == extension) {
                // Return resolved/absolute paths
                files.push_back(fs::absolute(entry.path()));
            }
        };

        try
        {
            if (recursive) {
                for (const auto& entry : fs::recursive_directory_iterator(directory)) {
                    add_file_if_match(entry);
                }
            } else {
                for (const auto& entry : fs::directory_iterator(directory)) {
                    add_file_if_match(entry);
                }
            }
        }
        catch(const std::exception& e)
        {
            std::cerr << "Error scanning directory " << directory.string() << ": " << e.what() << std::endl;
        }
        return files;
    }

} // namespace ImageToolkit

File: app/src/core/file_system_entries.py
import os
import shutil
import functools

from pathlib import Path


class FSETool:
    """
    A comprehensive tool for managing file system entries, including path 
    resolution, directory creation, file searching, and path manipulation.
    """
    # --- Utility Methods ---
    @staticmethod
    def path_contains(parent_path, child_path):
        """
        Check if parent_path contains child_path.
        Returns True if child_path is within parent_path or equal to it.
        """
        try:
            parent = Path(parent_path).resolve()
            child = Path(child_path).resolve()
            child.relative_to(parent)
            return True
        except (ValueError, TypeError):
            return False

    @staticmethod
    def prefix_create_directory(arg_id=0, kwarg_name='', is_filepath=False):
        """
        Decorator factory that returns a function to create a directory 
        based on a positional argument (arg_id) or keyword argument (kwarg_name).
        """
        def inner(*args, **kwargs):
            path = kwargs.get(kwarg_name, None)
            
            if path is None and len(args) > arg_id:
                path = args[arg_id]
            
            if path:
                directory = os.path.dirname(path) if is_filepath else path
                if not directory:
                    return True
                
                if os.path.exists(directory): 
                    return True
                try:
                    os.makedirs(directory, exist_ok=True)
                    print(f"Created directory: '{directory}'.")
                    return True
                except Exception as e:
                    print(f"ERROR: could not create directory '{directory}' with exception {type(e)}.")
                    raise
            return False
        return inner

    @staticmethod
    def ensure_absolute_paths(prefix_func=None, suffix_func=None):
        """
        Decorator factory to ensure paths are made absolute relative to 
        the system path.
        """
        def decorator(func):
            @functools.wraps(func)
            def wrapper(*args, **kwargs):
                if prefix_func is not None: prefix_func(*args, **kwargs)

                normalized_args = list(args)
                
                # Normalize positional arguments
                for id, arg in enumerate(normalized_args):
                    if isinstance(arg, str) and not os.path.isabs(arg):
                        if os.path.exists(arg):
                            normalized_args[id] = os.path.abspath(arg)
                        
                # Normalize keyword arguments
                for key, value in kwargs.items():
                    if isinstance(value, str) and not os.path.isabs(value) and os.path.exists(value):
                        kwargs[key] = os.path.abspath(value)

                result = func(*tuple(normalized_args), **kwargs)

                if suffix_func is not None: suffix_func(*normalized_args, **kwargs)
                
                return result
            return wrapper
        return decorator

    # --- Core File System Methods ---

    @staticmethod
    @ensure_absolute_paths()
    def get_files_by_extension(directory, extension, recursive=False):
        """
        Get all files with specific extension in directory.
        
        Args:
            directory: Path to directory
            extension: File extension (e.g., 'txt', '.jpeg')
            recursive: Whether to search subdirectories
        
        Returns:
            List of absolute file paths (strings).
        """
        path = Path(directory)
        
        if not extension.startswith('.'):
            extension = '.' + extension
        
        if recursive:
            pattern = f'**/*{extension}'
        else:
            pattern = f'*{extension}'
        
        files = [str(f.resolve()) for f in path.glob(pattern) if f.is_file()]
        return files


class FileDeleter:
    """
    A class dedicated to safely deleting files and directories, 
    leveraging FSETool for path resolution.
    """

    @staticmethod
    @FSETool.ensure_absolute_paths()
    def delete_path(path_to_delete: str) -> bool:
        """Deletes a file or directory recursively."""
        if not os.path.exists(path_to_delete):
            print(f"WARNING: specified path does not exist - did not delete '{path_to_delete}'.")
            return False
    
        if os.path.isdir(path_to_delete):
            try:
                shutil.rmtree(path_to_delete)
                print(f"Deleted directory: '{path_to_delete}'.")
            except OSError as e:
                print(f"ERROR: Could not delete directory {path_to_delete}. Reason: {e}")
                return False
        elif os.path.isfile(path_to_delete):
            try:
                os.remove(path_to_delete)
                print(f"Deleted file: '{path_to_delete}'.")
            except OSError as e:
                print(f"ERROR: Could not delete file {path_to_delete}. Reason: {e}")
                return False
        return True

    @staticmethod
    @FSETool.ensure_absolute_paths()
    def delete_files_by_extensions(directory: str, extensions: list[str]) -> int:
        """
        Recursively delete files with extension(s) in directory and all subdirectories.
        """
        path = Path(directory)
        deleted_count = 0
        
        for extension in extensions:
            if not extension.startswith('.'):
                extension = '.' + extension
            
            for file_path in path.rglob(f'*{extension}'):
                if file_path.is_file():
                    try:
                        file_path.unlink()
                        print(f"Deleted: {file_path}")
                        deleted_count += 1
                    except OSError as e:
                        print(f"ERROR: Could not delete file {file_path}. Reason: {e}")
        
        print(f"Deleted {deleted_count} files recursively.")
        return deleted_count
